File: examples/simple_chat.exs
# Port of examples/simple_chat.rs — interactive chat loop
#
# Usage:
#   mix run examples/simple_chat.exs --model path/to/model.gguf
#
# NOTE: ModelParams/ContextParams/SamplerChain are not yet exposed through
# the bindings. This example shows the target API shape using greedy
# sampling as a stand-in for the sampler chain.

defmodule SimpleChat do
  alias RustyLlama.{Model, Context, Sequence}

  def main(args) do
    {opts, _, _} =
      OptionParser.parse(args,
        strict: [model: :string, context: :integer, ngl: :integer],
        aliases: [m: :model, c: :context]
      )

    model_path = opts[:model] || raise "missing --model"
    _context_size = opts[:context] || 2048
    _ngl = opts[:ngl] || 99

    # Initialize the model
    # TODO: pass ModelParams with n_gpu_layers once params are exposed
    {:ok, model} = Model.load_from_file(model_path, %{})

    # Initialize the context
    # TODO: pass ContextParams with n_ctx/n_batch once params are exposed
    {:ok, ctx} = Context.new(model, %{})

    seq = Context.sequence(ctx)

    chat_loop(model, seq, [])
  end

  defp chat_loop(model, seq, messages) do
    input = IO.gets("") |> String.trim()

    if input == "" do
      :ok
    else
      messages = messages ++ [{:user, input}]

      prompt = format_messages(messages)
      is_first = Sequence.is_empty(seq)

      tokens = Model.tokenize(model, prompt, is_first, true)
      Sequence.extend(seq, tokens)

      IO.puts("")

      {response, _} = generate_response(model, seq)

      IO.puts("")

      messages = messages ++ [{:assistant, response}]
      chat_loop(model, seq, messages)
    end
  end

  defp generate_response(model, seq, response \\ "") do
    logits = Sequence.logits(seq)

    # Greedy argmax as stand-in for SamplerChain (min_p + temp + dist)
    {token, _score} =
      logits
      |> Enum.with_index()
      |> Enum.max_by(fn {score, _idx} -> score end)
      |> then(fn {score, idx} -> {idx, score} end)

    if Model.is_eog(model, token) do
      {response, seq}
    else
      piece = Model.token_to_piece(model, token)
      IO.write(piece)
      Sequence.push(seq, token)

      response = response <> piece

      if String.contains?(piece, "\n") do
        {response, seq}
      else
        generate_response(model, seq, response)
      end
    end
  end

  defp format_messages(messages) do
    formatted =
      Enum.map_join(messages, "\n", fn
        {:user, text} -> "user: #{text}"
        {:assistant, text} -> "assistant: #{text}"
      end)

    result = formatted <> "\nassistant:"
    IO.puts(result)
    result
  end
end

SimpleChat.main(System.argv())

File: examples/simple_chat.ts
// Port of examples/simple_chat.rs — interactive chat loop
//
// Usage:
//   node examples/simple_chat.js --model path/to/model.gguf
//
// Build the NAPI addon first:
//   cd packages/node && npm run build

import { Model, Context, type Sequence } from "../packages/node";
import * as readline from "node:readline";
import { parseArgs } from "node:util";

const { values } = parseArgs({
  options: {
    model: { type: "string", short: "m" },
    context: { type: "string", default: "2048" },
    ngl: { type: "string", default: "99" },
  },
});

const modelPath = values.model;
if (!modelPath) throw new Error("missing --model");

const ngl = Number(values.ngl ?? "99");
const nCtx = Number(values.context ?? "2048");

const model = Model.loadFromFile(modelPath, ngl);

const ctx = Context.new(model, nCtx, 512);

const seq = ctx.sequence()!;

interface Message {
  role: "user" | "assistant";
  content: string;
}

const messages: Message[] = [];

function formatMessages(msgs: Message[]): string {
  const formatted = msgs
    .map((m) => `${m.role}: ${m.content}`)
    .join("\n");
  const result = `${formatted}\nassistant:`;
  console.log(result);
  return result;
}

function generateResponse(): string {
  let response = "";

  while (true) {
    const logits = seq.logits();
    if (!logits) break;

    let bestToken = 0;
    let bestScore = -Infinity;
    for (let j = 0; j < logits.length; j++) {
      if (logits[j] > bestScore) {
        bestScore = logits[j];
        bestToken = j;
      }
    }

    if (model.isEog(bestToken)) break;

    const piece = model.tokenToPiece(bestToken);
    process.stdout.write(piece);
    response += piece;

    seq.push(bestToken);

    if (piece.includes("\n")) break;
  }

  return response;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function prompt() {
  rl.question("", (input) => {
    if (!input.trim()) {
      rl.close();
      return;
    }

    messages.push({ role: "user", content: input.trim() });

    const promptText = formatMessages(messages);
    const isFirst = seq.isEmpty();

    const tokens = model.tokenize(promptText, isFirst, true);
    seq.extend(tokens);

    console.log();
    const response = generateResponse();
    console.log();

    messages.push({ role: "assistant", content: response });
    prompt();
  });
}

prompt();
